scripts/fbi/hate_crime/download_publication_data.py

```diff
@@ -0,0 +1,363 @@
+                        _STATUS_DICT['cur_year']].append(table_xls_url)
+                    with open(
+                            os.path.join(_STATUS_DICT['output_dir'],
+                                         'url_list.json'), 'w') as fp:
+                        json.dump(_STATUS_DICT['download_urls'], fp, indent=2)
+                    # fetch and store urls
+                    r = requests.get(table_xls_url)
+                    output_path = os.path.join(_STATUS_DICT['output_dir'],
+                                               f"{_STATUS_DICT['cur_year']}")
+                    with open(os.path.join(output_path, f'table_{table_i}.xls'),
+                              'wb') as fp:
+                        fp.write(r.content)
+
+            time.sleep(0.5)
+
+
+def scrape_yearwise():
+    """The function controls the scrape flow and calls relevant functions for each stage.
+    Response from each stage is stored in the status_dict to be used for the next stage.
+    """
+    global _STATUS_DICT
+
+    session = requests.Session()
+    session.headers.update(_HEADERS)
+
+    _STATUS_DICT['session'] = session
+
+    _STATUS_DICT['output_dir'] = os.path.expanduser(_STATUS_DICT['output_dir'])
+    if not os.path.exists(_STATUS_DICT['output_dir']):
+        os.makedirs(_STATUS_DICT['output_dir'], exist_ok=True)
+
+    _STATUS_DICT['cache_dir'] = os.path.expanduser(_STATUS_DICT['cache_dir'])
+    if not os.path.exists(_STATUS_DICT['cache_dir']):
+        os.makedirs(_STATUS_DICT['cache_dir'], exist_ok=True)
+    if _STATUS_DICT['force_fetch'] or not os.path.isfile(
+            os.path.join(_STATUS_DICT['cache_dir'], 'base.html')):
+        base_html, base_soup = request_html_soup(session,
+                                                 _STATUS_DICT['base_url'])
+        with open(os.path.join(_STATUS_DICT['cache_dir'], 'base.html'),
+                  'w') as fp:
+            fp.write(base_html)
+    else:
+        with open(os.path.join(_STATUS_DICT['cache_dir'], 'base.html'),
+                  'r') as fp:
+            base_html = fp.read()
+            base_soup = BeautifulSoup(base_html, features="html.parser")
+
+    for cur_year in _STATUS_DICT['year_list']:
+        _STATUS_DICT['cur_year'] = cur_year
+        year_url = _find_year_url(base_soup, cur_year)
+        if year_url:
+            print(year_url)
+            _STATUS_DICT['cur_year_url'] = year_url
+            output_path = os.path.join(_STATUS_DICT['output_dir'],
+                                       f"{cur_year}")
+            if not os.path.exists(output_path):
+                os.makedirs(output_path, exist_ok=True)
+            _STATUS_DICT['download_urls'][cur_year] = []
+
+            access_url = _find_access_tables_url()
+
+            if access_url:
+                print(access_url)
+                _find_table_html_url()
+                _find_table_xls_url()
+            else:
+                print('No access tables url')
```

Comment:
could add context, `ERROR:` is this is an error as well as year, base url, etc.

Reply:
fixed.